### PR TITLE
Post: Improve how to run OpenNMS Horizon on Docker

### PR DESCRIPTION
**Pull request by indigo423**
Add blog post about the latest improvements how to run OpenNMS Horizon in container environment with Docker using the officially hosted images.

### CODE REVIEW

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+title: Running OpenNMS Horizon in Docker
+date: 21:48 2018-01-09
+author: Ronny Trommer
+body_classes: header-lite fullwidth blogstyling
+taxonomy:
+    category: docker
+    tag: [docker]
+---
+
+Running applications in containers provide a lot benefits and it's not just a hype.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: Running applications in containers provides many benefits, and it's not just hype.

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+Higher velocity to maintain changes while keeping a service available, scaling software and teams up to abstracting your infrastructure to use your expensive hardware more efficiently - all this is the fuzz about running applications in containers.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: s/fuzz/buzz/
Unclear what you mean by "scaling software and teams up to abstracting your infrastructure".

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+You should know OpenNMS is currently not built to run in all components in separated and isolated processes.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: s/in all components/with all components/

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+But you can gain some benefits running it as a legacy application in a container.
+There are already changes in the architecture to continuously improve this issue, but it will take time and I want to use it NOW&trade;.
+The technology is available and this article describes how to run OpenNMS Horizon in Docker and is an article more or less for the brave ones.
+
+Make yourself familiar how Docker works and data is persisted, there are Dragons!
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: familiar with how Docker works

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+If you run something in production without knowing, you get in trouble.
+The [Documentation](https://docs.docker.com/get-started/) is really good and worth reading and yes this is a DISCLAIMER :)
+
+To get the possibility to run OpenNMS Horizon in Docker we need container images which we provide on DockerHub.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: container images are needed, which we provide on DockerHub.

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+You will find images for [Horizon](https://hub.docker.com/r/opennms/horizon-core-web/) and [Minion](https://hub.docker.com/r/opennms/minion/).
+Images are built automatically, on  [CircleCI](https://circleci.com) for [Horizon](https://circleci.com/gh/opennms-forge/docker-horizon-core-web) and [Minion](https://circleci.com/gh/opennms-forge/docker-minion).
+The images based on a latest vanilla CentOS 7 and is installed from the official published RPMS.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: The images are based on... and are installed from... RPMs.

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+Related `Dockerfile` for [Horizon](https://github.com/opennms-forge/docker-horizon-core-web) and [Minion](https://github.com/opennms-forge/docker-minion) can be found on GitHub.
+
+The workflow to build and maintain these images is refactored and streamlined to reduce mainenance.
+Build and publishing workflow is now more transparent within the GitHub repository and allows feedback through contribution.
+
+Additionally there are some tips you need to know when running a Java applications in a Docker container.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: a Java application

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+## Building images with CircleCI
+
+First of all we build the Container Images with CircleCI which gives us the possibility to decentralise the build- and publishing process for container images.
+The whole build and publishing process is described in the GitHub repository in the `.circleci` directory.
+When you fork the project, you get the whole workflow for free and you can start immediately to modify and publish your own images as you like.
+
+## What means horizon-core-web
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: What does horizon-core-web mean
or
What horizon-core-web means

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+
+The provided image has just the bare minimum installed to run the core services for monitoring and the web application.
+If you want to use more advanced features like SNMP Hardware Asset adapter, RANCID integration or the JIRA ticketer integration plugins you have to install the RPMS on your own.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: RANCID integration, or the JIRA.... plugins, you have to...

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+The easiest way is to derive from the existing image which is described later in this article.
+It is recommended to build a service stack with _Docker Compose_ using an [official PostgreSQL](https://hub.docker.com/_/postgres/) data base as container image.
+
+## Provide an overlay etc-directory
+
+Updating from one OpenNMS Horizon version to another is not so easy and you need to merge a lot configuration files.
+
+The main issues which makes it difficult are:
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: issues which make it difficult

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+* Files which can be edited through the Web UI and are persisted in ${OPENNMS_HOME}/etc directory, this makes it hard to version control files which need to be edited on the file system and the running container depends on the file system.
+* Hard coded library version dependencies in Karaf properties which are stored in ${OPENNMS_HOME}/etc, e.g. `${OPENNMS_HOME}/etc/custom.properties` directory. Even if you never modify this files, you have to deal with them during an update.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: modify these files

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+* The Karaf cache is used to persist topology history and the GraphML topology
+
+To address the configuration management issues we provide an `etc-overlay` directory as convention.
+It allows to just maintain the configuration files you have touched without caring about the ones you don't have touched.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: about the ones you haven't touched.

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+You will run into similar problems if you try to use a configuration management tool or try to version your startup configuration of OpenNMS Horizon as well.
+
+The following mount command in your OpenNMS Horizon service definition allows you to use a local directory for your custom configurations on your Docker host system:
+
+```
+volumes:
+  - ./etc-overlay:/opt/opennms-etc-overlay
+```
+
+ATTENTION: You have to be aware, about configuration files changed in the Web UI and which are still persisted on the files system.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: s/files system/filesystem/

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+           This files need to be maintained, some examples are users.xml, groups.xml, snmp-config.xml, surveillance-views.xml, etc..
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: These files need to be maintained. Some examples are...

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+           Ideally this configuration files should be in the database but there is currently not really another useful workaround for this.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: s/this configuration/these configuration/
be in the database,

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+           You have to make sure your `${OPENNMS_HOME}/etc` directory gets persisted in a local volume.
+           The Karaf cache directory is by default not persisted and gets lost on restart.
+           If you work with the Topology UI ensure the `${OPENNMS_HOME}/data` gets persisted in a local volume as well, you have to deal with this during updates!
+
+## Image Tags
+
+We have cleaned up the image tags to provide the following use cases:
+
+* specific stable release version, e.g. `21.0.2`: Use this tag if you have a specific version dependencies in your service stack and you want full control for upgrades
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: s/a specific/specific/

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+* floating stable release tag `latest`: Use this tag when you want to run always the latest stable release, update the container image to latest stable release with `docker-compose pull && docker-compose stop opennms && docker-compose up -d`
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: s/run always/always run/
...release. Update the container...

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+* floating daily snapshot tag `bleeding`: This is an image built on a daily base from the latest `develop` branch and is the release candidate for the next major release of horizon
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: s/daily base/daily basis/

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+* feature tag `drift`: This is a special tag for daily snapshot built if you want to test the features which are currently developed in [Project Drift](https://wiki.opennms.org/wiki/DevProjects/Drift). This tag will be removed when this feature is merged in a released version of Horizon.
+
+## Allow Custom Health checks
+
+In previous container image versions the Health check where hard coded.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: s/where/were/

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+## Service Dependencies
+
+The old OpenNMS Horizon image had a sleep waiting for the PostgreSQL database.
+With Docker Compose version 2.1 and later it is possible to wait for a service condition.
+It makes the wait obsolete and OpenNMS Horizon is started when the PostgreSQL database is in state "healthy".
+
+## Java Process and PID 1
+
+Processes with ID 1 have special functionalities related to signal handling and cleaning up zombie processes.
+Using a current the example `docker-compose.yml` uses now the built-in `init: true` option, which runs an tiny init inside the container and prevents the Java process to get PID 1.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: For example, `docker-compose.yml` now uses the built-in `init: true` option, which runs a tiny init inside the container and prevents the Java process from getting PID 1.

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 21]
```diff
@@ -0,0 +1,171 @@
...
+The init option requires to use a Docker-Compose in version 2.2 or later.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: s/to use/using/
s/in version/version/

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 22]
```diff
@@ -0,0 +1,171 @@
...
+
+## Derive from the Horizon Image
+
+If you want to build and deploy your own image, just [fork](https://help.github.com/articles/fork-a-repo/) the [project](https://github.com/opennms-forge/docker-horizon-core-web).
+Create your custom Dockerfile for example named `MyDockerfile` with the following content:
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: ...Dockerfile, for example..., with the following content:

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 23]
```diff
@@ -0,0 +1,171 @@
...
+```
+
+If you want to do this with CircleCI automatically you have to provide your DockerHub credentials as [project environment variable](https://circleci.com/docs/2.0/env-vars/#adding-environment-variables-in-the-app).
+The variables which are used in the config are `DOCKERHUB_LOGIN` and `DOCKERHUB_PASS`.
+
+## Container and Java Memory Management with Limits
+
+Running a Java application in a container has a few things to consider.
+To control and isolate resource usage of processes the Kernel feature _cgroups (Control Groups)_ is used.
+If you do nothing, the _Maximum Heap Size_ and _Initial Heap Size_ is calculated based on the memory of the Docker host system and can lead into problems when a memory limit on the container is set.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: s/lead into problems/lead to problems/

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 24]
```diff
@@ -0,0 +1,171 @@
...
+Running a JVM on a server-class hardware means 1/64th of the machines physical memory is used as _Initial Heap Size_.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: ...on server-class...
It's unclear what you mean by "server-class hardware," or how that impacts memory calculations. By context, it appears you mean "a machine with 64 GB of RAM."

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 25]
```diff
@@ -0,0 +1,171 @@
...
+For the _Maximum Heap Size_ less than 1/4th of the physical memory or 1GB is used.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: 'th' is not needed for 1/4.

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md [hunk 26]
```diff
@@ -0,0 +1,171 @@
...
+If you want to go with option 2 you have to set the following Java options:
+
+```
+JAVA_OPTS=-XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap
+```
+
+Previous Docker images used the opennms startup bash script with some predefined ulimit settings.
+The current version runs the native Java command in the [entrypoint script](https://github.com/opennms-forge/docker-horizon-core-web/blob/master/docker-entrypoint.sh#L86) and the `JAVA_OPTS` environment variable allows you to configure the JVM.
+The stack size is 8192 kb by default and the open file limit is `1048576`.
+We use now the defaults and which makes it more transparent.
```
Comment (UberPinguin) [inline user/pages/05.blog/2018-01-09-docker-improvements/item.md]: We now use the defaults, which makes it more transparent.

**Review by UberPinguin — COMMENTED** [review]
Some grammar and punctuation suggestions, and questions about clarity.

File: user/pages/05.blog/2018-01-09-docker-improvements/item.md (hunk 24 above)
Comment (indigo423) [reply user/pages/05.blog/2018-01-09-docker-improvements/item.md]: I've added a link what server-class in JVM means. Does it make sense now?

**Comment by indigo423** [comment]
@UberPinguin thank you so much for feedback!

🤘

**Pull request merged**